### DISCUSSION TITLE
如何让回答更加精准

### DISCUSSION THREAD

**Discussion by WilliamChen-luckbob** [chatchat-space/Langchain-Chatchat#1565] [Q&A]
比如一个场景，同一个类型的不同型号机器，它们拥有相似的特性，但是会有细节差异。

将其技术文档扫入知识库，在提问时对某种通用特性进行提问，得到的答案是比较靠谱的。

然而会有一个问题，如果对某个型号的特性进行具体提问，往往无法获得靠谱的回答。

下面是我的测试场景：
机型说明书，型号分别为classic M01, classic M02, classic M03, classic M04, classic M05

我看过默认unstructured的文档切片逻辑，切出来的语句块会被装载成如下形式
文档切分示例：page_content="实际的切分出来的原文部分" metadata={'source': '文档路径'}

此数据将会被转为向量，并丢到向量库中
在实际查询时，指定型号将会把其它的型号也加载出来，但是指定的型号会排在第一位，而后续是其它型号的相似语句块。
比如
```
问：classic M01有哪些特性

预期：
classic M01 的功能特性如下：
A特性XXXXXX
B特性XXXXXX
C特性XXXXXX
D特性XXXXXX

实际：
classic M01 的功能特性如下：
classic M01的A特性XXXXXX
classic M04的A特性XXXXXX
classic M01的C特性XXXXXX
classic M02的B特性XXXXXX
classic M05的D特性XXXXXX

-------------------------------------------------------------------------------------------------

匹配1
功能特性A是XXXX
出处：文档classic M01

匹配2
功能特性A是XXXX
出处：文档classic M04

匹配3
功能特性C是XXXX
出处：文档classic M01

匹配4
功能特性B是XXXX
出处：文档classic M02

匹配5
功能特性D是XXXX
出处：文档classic M05
```

后面我尝试修改了一下，在导入文本时，存一个model字段在metadata中
文档切分示例：page_content="实际的切分出来的原文部分" metadata={'source': '文档路径', 'model': '实际的型号'}

同样的提问，效果更差了，比如下面的例子，甚至连指定的模型的文件都找不到

```
问：classic M01有哪些特性

预期：
classic M01 的功能特性如下：
A特性XXXXXX
B特性XXXXXX
C特性XXXXXX
D特性XXXXXX

实际：
classic M01 的功能特性如下：
classic M03的A特性XXXXXX
classic M04的A特性XXXXXX
classic M03的C特性XXXXXX
classic M02的B特性XXXXXX
classic M05的D特性XXXXXX

-------------------------------------------------------------------------------------------------

匹配1
功能特性A是XXXX
出处：文档classic M03

匹配2
功能特性A是XXXX
出处：文档classic M04

匹配3
功能特性C是XXXX
出处：文档classic M03

匹配4
功能特性B是XXXX
出处：文档classic M02

匹配5
功能特性D是XXXX
出处：文档classic M05
```

**Comment by WilliamChen-luckbob** [comment]
问题已经解决，需要自行fork项目修改后搭建，项目源码不支持使用metadata进行索引，但是引用的langchain方法是有带索引的重载的，替换方法并想办法将型号带入，即可精准匹配，技术文档的准确率测试超过80%

**Reply by hihanyi** [reply]
大佬可以具体说说怎么实现的吗

**Reply by WilliamChen-luckbob** [reply]
可以参考#1943，可以保证在指定的型号中获取文本块，至于获取的精准程度，跟你的text splitter大有关系
